test/ruby/test_shapes.rb

```diff
@@ -226,6 +226,52 @@ def initialize
     end;
   end
 
+  def test_run_out_of_shape_generic_ivar_set
+    assert_separately([], "#{<<~"begin;"}\n#{<<~'end;'}")
+    begin;
+      class TooComplex < Hash
+      end
+
+      # Try to run out of shapes
+      o = Object.new
+      i = 0
+      while RubyVM::Shape.shapes_available > 0
+        o.instance_variable_set(:"@i#{i}", 1)
+        i += 1
+      end
+
+      tc = TooComplex.new
+      tc.instance_variable_set(:@a, 1)
+      tc.instance_variable_set(:@b, 2)
+    end;
+  end
```

Comment:
I found another crash on our CI, so I pushed a second commit to handle it.

I'm running yet another build to see if there's other cases we don't handle.